File: bench/btree.rb
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# contributed by Jesse Millikan
# Modified by Wesley Moxam
# Modified by Scott Leggett
# *reset*
# slightly modified
def item_check(left, right)
    if left
        1 + item_check(*left) + item_check(*right)
    else
        1
    end
end

def bottom_up_tree(depth)
    if depth > 0
        depth -= 1
        [bottom_up_tree(depth), bottom_up_tree(depth)]
    else
        [nil, nil]
    end
end

max_depth = 15
min_depth = 4

max_depth = [min_depth + 2, max_depth].max

stretch_depth = max_depth + 1
stretch_tree = bottom_up_tree(stretch_depth)

puts "stretch tree of depth #{stretch_depth}\t check: #{item_check(*stretch_tree)}"
stretch_tree = nil

long_lived_tree = bottom_up_tree(max_depth)

min_depth.step(max_depth, 2) do |depth|
  iterations = 2**(max_depth - depth + min_depth)

  check = 0

  (1..iterations).each do |i|
    check += item_check(*bottom_up_tree(depth))
  end

  puts "#{iterations}\t trees of depth #{depth}\t check: #{check}"
end

puts "long lived tree of depth #{max_depth}\t check: #{item_check(*long_lived_tree)}"

File: bench/btree.js
function make_tree(depth){
    if (depth==0) {
        return [null,null]
    }
    depth -= 1
    return [make_tree(depth), make_tree(depth)]
}

function check_tree(node){
    let left=node[0]
    let right=node[1]
    if(left){
        return 1 + check_tree(left) + check_tree(right)
    }
    return 1
}

let min_depth = 4
let max_depth = 15
let stretch_depth = max_depth + 1
console.log(`stretch tree of depth ${stretch_depth}\t check:${check_tree(make_tree(stretch_depth))}`)

let long_lived_tree = make_tree(max_depth)

let iterations = 32768//2**15

let depth=min_depth

while(depth < stretch_depth){
    let check = 0
    for(let i=0;i<iterations;i++){
        check += check_tree(make_tree(depth))
    }

    console.log(`${iterations}\t trees of depth ${depth}\t check:${check}`)
    iterations /= 4
    depth+=2
}

console.log(`long lived tree of depth ${max_depth} \t check ${check_tree(long_lived_tree)}`)
